### PR TITLE
Refactor/backtest checks

### PR DESCRIPTION
**Pull request by LeoTafti**
### Summary

- Changed `RegressionModel.backtest()` `start` parameter type to match `ForecastingModel.backtest()` signature.
- Factored out some code used in both forecasting and regression model to perform sanity checks on arguments and fixed a couple of bugs with these checks.
- Added a check on `start` to make sure it doesn't correspond to the first timestamp in the series, since this would leave no points in the training set and make it impossible to make a first prediction.
- Added a couple more unit tests for both forecasting and regression model `backtest()` methods.

### Things I'm unsure about

- I set the default value of `start` for `StandardRegressionModel` to be equal to `train_n_points`. I think that this is the most intuitive choice.
- I'm not sure which default value to give to `start` in the general case (ie. in `RegressionModel.backtest()`). I set it to 1, since that is the smallest possible index which makes sense.
- I put `_backtest_general_checks()` and `_backtest_convert_start()` in `utils.py` since I couldn't think of a better place to put code in common for both ForecastingModel and RegressionModel, but there might be a better place.

Ran tests locally, all green.

### CODE REVIEW

**Comment by LeoTafti** [comment]
Also what would think about giving a warning when `StandardRegressionModel.fit()` is called with `train_n_points` larger than the number of available training points in `training_series`?

**Comment by LeoTafti** [comment]
Fixes #188

File: darts/utils/utils.py [hunk 1]
```diff
@@ -110,9 +117,84 @@ def fit(self, a, b=0, c=0):
...
+    # check that trim_to_series and start together form a valid combination
+    trim_to_series = n.trim_to_series
+
+    if trim_to_series:
+        raise_if_not(start + series.freq() * forecast_horizon in series,
+                     '`start` timestamp is too late in the series to make any predictions with'
+                     '`trim_to_series` set to `True`.', logger)
+
+
+def _backtest_convert_start(start: Union[pd.Timestamp, float, int], training_series: TimeSeries) -> pd.Timestamp:
```
Comment (hrzn) [inline darts/utils/utils.py:194]: I think this method might also be useful outside of backtesting as well. You could rename it to reflect this. For instance to something like `get_ts_at_point(point: Union[pd.Timestamp, float, int], series: TimeSeries) -> pd.Timestamp`.

In fact, I think later on (not for this PR) we could even think of having such "fancy indexing" natively supported in `TimeSeries`, as well as in some `TimeSeries` methods such as `split_after`, `split_after`, `slice`, etc.

File: darts/models/standard_regression_model.py [hunk 2]
```diff
@@ -70,3 +70,24 @@ def predict(self, features: List[TimeSeries]):
         super().predict(features)
         y = self.model.predict(self._get_features_matrix_from_series(features))
         return TimeSeries(pd.DataFrame(y, index=features[0].time_index()), self.train_target.freq())
+
+    def backtest(self,
```
Comment (hrzn) [inline darts/models/standard_regression_model.py:74]: I don't think it's worth to rewrite the backtest() method here just to add the default `start = self.train_n_points`.

With this model, `self.train_n_points` can be set to `None`, in which case the model will just retrain on whatever is available. I think we could assume that this is the "default", and remove this backtest() implementation. We should however slightly change the `__init__` function so that the argument becomes `train_n_points: Optional[int]=None`.

File: darts/models/regression_model.py [hunk 3]
```diff
@@ -84,11 +84,41 @@ def predict(self, features: List[TimeSeries]) -> TimeSeries:
...
+        """Method to be overriden in subclass for model specific sanity checks"""
+        pass
+
+    @_with_sanity_checks("_backtest_sanity_checks", "_backtest_model_specific_sanity_checks")
     def backtest(self,
                  feature_series: Iterable[TimeSeries],
                  target_series: TimeSeries,
-                 start: pd.Timestamp,
-                 forecast_horizon: int,
+                 start: Union[pd.Timestamp, float, int] = 1,
```
Comment (hrzn) [inline darts/models/regression_model.py:120]: Not too sure about having a default value for this parameter. Or if we have one, we should be consistent with ForecastingModel and put 0.5. (although I think 0.7 on both models would be better, actually)

File: darts/tests/test_backtesting.py [hunk 4]
```diff
@@ -115,6 +140,24 @@ def test_backtest_regression(self):
...
+        # Using a too small start value
+        with self.assertRaises(ValueError):
+            StandardRegressionModel(15).backtest(features, target, start=0, forecast_horizon=3)
+
+        with self.assertRaises(ValueError):
+            StandardRegressionModel(15).backtest(features, target, start=0.01, forecast_horizon=3)
+
+        # Using StandardRegressionModel's start default value
+        pred = StandardRegressionModel(15).backtest(features, target, forecast_horizon=3)
+        self.assertEqual(r2_score(pred, target), 1.0)
```
Comment (hrzn) [inline darts/tests/test_backtesting.py:159]: 👍   for the tests :)

**Review by hrzn — COMMENTED** [review]
Nice work :)

File: darts/utils/utils.py (hunk 1 above)
Comment (LeoTafti) [reply darts/utils/utils.py:194]: I think so as well.

Being able to at least split a `TimeSeries` according to some ratio or index would indeed be really nice, and this method could help there.

File: darts/models/standard_regression_model.py (hunk 2 above)
Comment (LeoTafti) [reply darts/models/standard_regression_model.py:74]: > With this model, self.train_n_points can be set to None, in which case the model will just retrain on whatever is available. I think we could assume that this is the "default", and remove this backtest() implementation.

Indeed, `train_n_points` can totally be made Optional assuming this default behavior.

However I still think that it would be somewhat misleading to silently train on less than `train_n_points` when `train_n_points` has been explicitly specified. What do you think ?

File: darts/models/regression_model.py (hunk 3 above)
Comment (LeoTafti) [reply darts/models/regression_model.py:120]: I wasn't too sure either.
The problem I see with making it 0.5 (or 0.7) as in ForecastingModel is that in the both model don't "work" with the same input series:

- In the ForecastingModel case, the series `backtest()` is "working" on (ie. `training_series`) is typically the "whole" series dataset (e.g. from 2000-01-01 to 2020-01-01)
- Whereas RegressionModel uses  `features_series` which are mostly already forecasts of other models (and would e.g. go each go from **2010**-01-01 to 2020-01-01.

I think the typical use case of RegressionModel would thus have a `start` param pointing to the very beginning of `feature_series`, and not at 50% or 70%

**Comment by hrzn** [comment]
> Also what would think about giving a warning when `StandardRegressionModel.fit()` is called with `train_n_points` larger than the number of available training points in `training_series`?

That's a good idea

File: darts/models/standard_regression_model.py (hunk 2 above)
Comment (hrzn) [reply darts/models/standard_regression_model.py:74]: That's a good point. Like you suggest, we should probably display a warning when `train_n_points` has been specified and the model is fit on fewer points. Then in `backtest` the warning would also be displayed as it calls the fit() method.

File: darts/models/regression_model.py (hunk 3 above)
Comment (hrzn) [reply darts/models/regression_model.py:120]: Fair point, however there is still the notion of training and validation set. So I think it remains intuitive (at least to me) that you would train regression models on the first 70% of the data available (so from 2010 to 2017) and then validate on the recent past only. Training on 1 data point is a bit awkward. So I'd vote for putting 70% everywhere or, if you really don't like it, to drop this default value altogether from `RegressionModel.backtest()`.

Comment (LeoTafti) [reply darts/models/regression_model.py:120]: Fair enough, I think we should go for 70% by default everywhere then.

**Review by hrzn — APPROVED** [review]
Thanks @LeoTafti !
